### ISSUE TITLE
create readonly columns

### ISSUE THREAD

**Issue by pleerock** [typeorm/typeorm#440]
need to have ability to mark columns as readonly which will allow updating only on insertion.

**Comment by idchlife** [comment]
@Column({ readonly: true }) ?

**Comment by pleerock** [comment]
probably so

**Comment by 19majkel94** [comment]
Isn't `public readonly property: Type` not enough? You will get compile time errors then, unlike the decorator solution which will throw error on persisting in db.

**Comment by idchlife** [comment]
@19majkel94 I wonder, if there is possibility to make readonly (only in persist write) columns at database level?

**Comment by pleerock** [comment]
> Isn't public readonly property: Type not enough? You will get compile time errors then, unlike the decorator solution which will throw error on persisting in db.

don't think so. `readonly: true` is mostly for security reasons, for example if user sent from the client something that should not be simply updated in a db. `readonly property` is just compile-time check, it wont secure dynamic transformations. Or its easy to avoid by having a constructor with exist id and readonly property passed.

>  if there is possibility to make readonly (only in persist write) columns at database level?

haven't heard about such

**Comment by 19majkel94** [comment]
>  is just compile-time check, it wont secure dynamic transformations.

Yeah, it wont work with `Repository.merge` but with simple `entity.readonlyProp = "something"` it will won't let you to compile.

If you want dynamic checking, it can be done with custom `@Readonly` property decorator that will use `Proxy` and throw error on set. So then simple property assignment or merge will give you runtime error. But you can still do dirty hacks and create an literal object and set the prototype to the entity class prototype - persist will update it.

So maybe it can be done on entity level? Maybe a `@BeforeUpdate` decorator which will throw an error if  the property has been changed? But I don't know if `TypeORM` inject the data to the handler method. If not, the only option is `EntitySubscriber` which will get `updatedColumns` array on `UpdateEvent` 😉

**Comment by pleerock** [comment]
Don't confuse different purposes. 

`readonly property` is for compile-time safety.
`@Column({ readonly: true })` or separate `@Readonly` decorator are to make sure user do not update something that he should not.

> create an literal object and set the prototype to the entity class prototype - persist will update it.

yes you can also simply run `query("DROP DATABASE");` and everything will be dropped. You can also do self-injections. Such examples aren't relevant.

**Comment by pleerock** [comment]
This feature is implemented and will be released in next (`0.1.0-alpha.38`) version. Usage:

```ts
import {Entity} from "../../../../../src/decorator/entity/Entity";
import {Column} from "../../../../../src/decorator/columns/Column";
import {PrimaryColumn} from "../../../../../src/decorator/columns/PrimaryColumn";

@Entity()
export class Post {

    @PrimaryGeneratedColumn()
    id: number;

    @Column()
    title: string;

    @Column()
    text: string;

    @Column({ readonly: true })
    authorName: string;

}
```

**Comment by idchlife** [comment]
this is crazy cool!
